File: Python/PySparkExamples/src/examples/UDFExample.py
from pyspark.sql import SparkSession
from pyspark import StorageLevel, Row
from pyspark.sql.functions import col, lit, array_contains, udf, split
from pyspark.sql.types import StructType, StructField, StringType, IntegerType, ArrayType

# Creation of spark object
spark = SparkSession.builder \
    .master("local[1]") \
    .appName("PySpark Examples") \
    .getOrCreate()

columns = ["Seqno","Name"]
data = [("1", "john jones"),
        ("2", "tracey smith"),
        ("3", "amy sanders")]

df = spark.createDataFrame(data=data,schema=columns)
df.show(truncate=False)

#Create a Python Function
def convertCase(str):
    return str

#Convert a Python function to PySpark UDF
""" Converting function to UDF """
convertUDF = udf(lambda z: convertCase(z),StringType())

""" 
Note: The default type of the udf() is StringType hence, you can also write the above statement without return type.
Converting function to UDF 
StringType() is by default hence not required 
convertUDF = udf(lambda z: convertCase(z)) 
"""

#Using UDF with PySpark DataFrame select()
df.select(col("Seqno"), \
          convertUDF(col("Name")).alias("Name")) \
    .show(truncate=False)


#Using UDF with PySpark DataFrame withColumn()
def upperCase(str):
    return str

upperCaseUDF = udf(lambda z:upperCase(z),StringType())
df.withColumn("Curated Name", upperCaseUDF(col("Name"))) \
    .show(truncate=False)

#Registering PySpark UDF & use it on SQL
""" Using UDF on SQL """
spark.udf.register("convertUDF", convertCase,StringType())
df.createOrReplaceTempView("NAME_TABLE")
spark.sql("select Seqno, convertUDF(Name) as Name from NAME_TABLE") \
    .show(truncate=False)


#Creating UDF using annotation
@udf(returnType=StringType())
def upperCase(str):
    return str.upper()

df.withColumn("Curated Name", upperCase(col("Name"))) \
    .show(truncate=False)


#Execution order
""" 
No guarantee Name is not null will execute first
If convertUDF(Name) like '%John%' execute first then 
you will get runtime error
"""
spark.sql("select Seqno, convertUDF(Name) as Name from NAME_TABLE " + \
          "where Name is not null and convertUDF(Name) like '%John%'") \
    .show(truncate=False)

#Handling null check
""" null check """

columns = ["Seqno","Name"]
data = [("1", "john jones"),
        ("2", "tracey smith"),
        ("3", "amy sanders"),
        ('4',None)]

df2 = spark.createDataFrame(data=data,schema=columns)
df2.show(truncate=False)
df2.createOrReplaceTempView("NAME_TABLE2")

spark.sql("select convertUDF(Name) from NAME_TABLE2") \
    .show(truncate=False)


spark.udf.register("_nullsafeUDF", lambda str: convertCase(str) if not str is None else "" , StringType())

spark.sql("select _nullsafeUDF(Name) from NAME_TABLE2") \
    .show(truncate=False)

spark.sql("select Seqno, _nullsafeUDF(Name) as Name from NAME_TABLE2 " + \
          " where Name is not null and _nullsafeUDF(Name) like '%John%'") \
    .show(truncate=False)

File: Scala/Spark-Scala-Examples/src/main/scala/com/examples/UDFExample.scala
package com.examples

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.sql.functions.{col,asc,desc, udf}
import org.apache.spark.sql.functions.udf

object UDFExample extends App{

  val spark:SparkSession = SparkSession
    .builder()
    .master("local[1]")
    .appName("Spark-Scala-Examples")
    .getOrCreate()

  import spark.implicits._
  val columns = Seq("Seqno","Quote")
  val data = Seq(("1", "Be the change that you wish to see in the world"),
    ("2", "Everyone thinks of changing the world, but no one thinks of changing himself."),
    ("3", "The purpose of our lives is to be happy.")
  )
  val df = data.toDF(columns:_*)
  df.show(false)

  val convertCase =  (strQuote:String) => {
    val arr = strQuote.split(" ")
    arr.map(f=>  f.substring(0,1).toUpperCase + f.substring(1,f.length)).mkString(" ")
  }

  val convertUDF = udf(convertCase)


  //Using with DataFrame
  df.select(col("Seqno"),
    convertUDF(col("Quote")).as("Quote") ).show(false)

  //Registering Spark UDF to use it on SQL
  // Using it on SQL
  spark.udf.register("convertUDF", convertCase)
  df.createOrReplaceTempView("QUOTE_TABLE")
  spark.sql("select Seqno, convertUDF(Quote) from QUOTE_TABLE").show(false)

}
